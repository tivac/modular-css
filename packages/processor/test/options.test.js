"use strict";

const path = require("path");

const dedent   = require("dedent");
const namer    = require("@modular-css/test-utils/namer.js");
const relative = require("@modular-css/test-utils/relative.js");
const logs     = require("@modular-css/test-utils/logs.js");

const Processor = require("../processor.js");

const sync = (css) => css.append({ selector : "a" });

const async = (css) => (
    new Promise((resolve) =>
        setTimeout(() => {
            sync(css);

            resolve();
        }, 0)
    )
);

describe("/processor.js", () => {
    describe("options", () => {
        describe("cwd", () => {
            it("should use an absolute path", async () => {
                const cwd = path.resolve("./packages/processor/test/specimens/folder");

                const processor = new Processor({ cwd });

                const { id } = await processor.file("./folder.css");

                expect(processor.options.cwd).toBe(cwd);
                expect(id).toBe(require.resolve("./specimens/folder/folder.css"));
            });

            it("should accept a relative path but make it absolute", async () => {
                const cwd = "./packages/processor/test/specimens/folder";

                const processor = new Processor({ cwd });

                const { id } = await processor.file("./folder.css");

                expect(processor.options.cwd).toBe(path.resolve(cwd));
                expect(id).toBe(require.resolve("./specimens/folder/folder.css"));
            });
        });

        describe("namer", () => {
            it("should use a custom naming function", async () => {
                const processor = new Processor({
                    namer : (filename, selector) =>
                        `${relative([ filename ])[0].replace(/[\/\.]/g, "_")}_${selector}`,
                });

                const { exports : compositions, details } = await processor.string(
                    "./packages/processor/test/specimens/simple.css",
                    ".wooga { }"
                );

                expect(compositions).toMatchSnapshot();
                expect(details.classes).toMatchSnapshot();
                expect(details.processed.root.toResult().css).toMatchSnapshot();
            });

            it("should require a namer if a string is passed", async () => {
                const processor = new Processor({
                    namer : "@modular-css/shortnames",
                });

                const result = await processor.string(
                    "./test/specimens/simple.css",
                    ".wooga { }"
                );

                expect(result.exports).toMatchSnapshot();
            });

            it("should use the default naming function if a non-function is passed", async () => {
                const processor = new Processor({
                    namer : false,
                });

                const result = await processor.string(
                    "./packages/processor/test/specimens/simple.css",
                    ".wooga { }"
                );

                expect(result.exports).toMatchSnapshot();
            });
        });

        describe("map", () => {
            it("should generate source maps", async () => {
                const processor = new Processor({
                        namer,
                        map : true,
                    });

                await processor.file("./packages/processor/test/specimens/start.css");

                const { css } = await processor.output({
                    from : "packages/processor/test/specimens/rewrite.css",
                    to   : "out.css",
                });

                expect(css).toMatchSnapshot();
            });

            it("should generate external source maps", async () => {
                const processor = new Processor({
                        namer,
                        map : {
                            internal : false,
                        },
                    });

                await processor.file("./packages/processor/test/specimens/start.css");

                const { css } = await processor.output({
                    from : "packages/processor/test/specimens/rewrite.css",
                    to   : "out.css",
                });

                expect(css).toMatchSnapshot();
            });
        });

        describe("exportGlobals", () => {
            it("should not export :global values when exportGlobals is false", async () => {
                const processor = new Processor({
                        exportGlobals : false,
                    });

                const { exports } = await processor.string(
                    "./exportGlobals.css",
                    dedent(`
                        :global(.a) {}
                        .b {}
                    `)
                );

                expect(exports).toMatchSnapshot();
            });
        });

        describe("rewrite", () => {
            it("should rewrite url() references by default", async () => {
                const processor = new Processor();

                await processor.string(
                    "packages/processor/test/specimens/rewrite.css",
                    dedent(`
                        .a {
                            background: url("img.png");
                        }
                    `)
                );

                const { css } = await processor.output({
                    from : "packages/processor/test/specimens/rewrite.css",
                    to   : "./packages/processor/test/output/rewrite.css",
                });

                expect(css).toMatchSnapshot();
            });

            it("should not rewrite url() references when falsey", async () => {
                const processor = new Processor({ rewrite : false });

                await processor.string(
                    "packages/processor/test/specimens/rewrite.css",
                    dedent(`
                        .a {
                            background: url("img.png");
                        }
                    `)
                );

                const { css } = await processor.output({
                    from : "packages/processor/test/specimens/rewrite.css",
                    to   : "./packages/processor/test/output/rewrite.css",
                });

                expect(css).toMatchSnapshot();
            });

            it("should pass through to postcss-url as config", async () => {
                const processor = new Processor({
                    rewrite : {
                        url : "inline",
                    },
                });

                await processor.string(
                    "packages/processor/test/specimens/rewrite.css",
                    dedent(`
                        .a {
                            background: url("img.png");
                        }
                    `)
                );

                const { css } = await processor.output({
                    from : "packages/processor/test/specimens/rewrite.css",
                    to   : "./packages/processor/test/output/rewrite.css",
                });

                expect(css).toMatchSnapshot();
            });
        });

        describe("postcss options", () => {
            it("should support custom parsers", async () => {
                const parser = require("sugarss");

                const processor = new Processor({
                    postcss : {
                        parser,
                    },
                });

                await processor.string(
                    "packages/processor/test/specimens/parser.css",
                    dedent(`
                        .a
                            color: blue
                    `)
                );

                const { css } = await processor.output({
                    from : "packages/processor/test/specimens/parser.css",
                    to   : "./packages/processor/test/output/parser.css",
                });

                expect(css).toMatchSnapshot();
            });
        });

        describe("lifecycle options", () => {
            describe("before", () => {
                it("should run sync postcss plugins before processing", async () => {
                    const processor = new Processor({
                        namer,
                        before : [ sync ],
                    });

                    await processor.string(
                        "packages/processor/test/specimens/sync-before.css",
                        ""
                    );

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/sync-before.css",
                    });

                    expect(css).toMatchSnapshot();
                });

                it("should run async postcss plugins before processing", async () => {
                    const processor = new Processor({
                        namer,
                        before : [ async ],
                    });

                    await processor.string(
                        "packages/processor/test/specimens/async-before.css",
                        ""
                    );

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/sync-before.css",
                    });

                    expect(css).toMatchSnapshot();
                });
            });

            describe("processing", () => {
                it("should run sync postcss plugins processing processing", async () => {
                    const processor = new Processor({
                        namer,
                        processing : [ sync ],
                    });

                    await processor.string(
                        "packages/processor/test/specimens/sync-processing.css",
                        ""
                    );

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/sync-processing.css",
                    });

                    expect(css).toMatchSnapshot();
                });

                it("should run async postcss plugins processing processing", async () => {
                    const processor = new Processor({
                        namer,
                        processing : [ async ],
                    });

                    await processor.string(
                        "packages/processor/test/specimens/async-processing.css",
                        ""
                    );

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/sync-processing.css",
                    });

                    expect(css).toMatchSnapshot();
                });

                it("should include exports from 'modular-css-export' modules", async () => {
                    const processor = new Processor({
                        namer,
                        processing : [ (css, { messages }) => {
                            messages.push({
                                plugin  : "modular-css-exporter",
                                exports : {
                                    a : true,
                                    b : false,
                                },
                            });
                        } ],
                    });

                    const { details } = await processor.string(
                        "packages/processor/test/specimens/async-processing.css",
                        ""
                    );

                    expect(details.exported).toMatchSnapshot();
                });
            });

            describe("after", () => {
                it("should use postcss-url by default", async () => {
                    const processor = new Processor();

                    await processor.file("./packages/processor/test/specimens/relative.css");

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/rewrite.css",
                        to   : "./packages/processor/test/output/relative.css",
                    });

                    expect(css).toMatchSnapshot();
                });

                it("should run sync postcss plugins", async () => {
                    const processor = new Processor({
                        namer,
                        after : [ sync ],
                    });

                    await processor.file("./packages/processor/test/specimens/relative.css");

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/rewrite.css",
                        to   : "./packages/processor/test/output/relative.css",
                    });

                    expect(css).toMatchSnapshot();
                });

                it("should run async postcss plugins", async () => {
                    const processor = new Processor({
                        namer,
                        after : [ async ],
                    });

                    await processor.file("./packages/processor/test/specimens/relative.css");

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/rewrite.css",
                        to   : "./packages/processor/test/output/relative.css",
                    });

                    expect(css).toMatchSnapshot();
                });
            });

            describe("done", () => {
                it("should run sync postcss plugins done processing", async () => {
                    const processor = new Processor({
                        namer,
                        done : [ sync ],
                    });

                    await processor.string(
                        "packages/processor/test/specimens/sync-done.css",
                        ""
                    );

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/sync-done.css",
                    });

                    expect(css).toMatchSnapshot();
                });

                it("should run async postcss plugins done processing", async () => {
                    const processor = new Processor({
                        namer,
                        done : [ async ],
                    });

                    await processor.string(
                        "packages/processor/test/specimens/async-done.css",
                        ""
                    );

                    const { css } = await processor.output({
                        from : "packages/processor/test/specimens/async-done.css",
                    });

                    expect(css).toMatchSnapshot();
                });
            });

            describe("verbose", () => {
                it("should output debugging messages when verbose mode is enabled", async () => {
                    const { logSnapshot } = logs();

                    const processor = new Processor({
                        namer,
                        verbose : true,
                    });

                    await processor.file("./packages/processor/test/specimens/start.css");
                    await processor.string(
                        "packages/processor/test/specimens/string.css",
                        ".foo { color: fuschia; }"
                    );

                    await processor.output();

                    logSnapshot();
                });
            });

            describe("dupewarn", () => {
                // const fn = cased ? it.skip : it;

                it("should warn on potentially duplicate file paths", async () => {
                    const { logSnapshot } = logs("warn");

                    const processor = new Processor({
                        namer,
                    });

                    await processor.string("packages/processor/test/specimens/start.css", ".start { color: red; }");
                    await processor.string("packages/processor/test/specimens/START.css", ".start { color: red; }");

                    logSnapshot();
                });

                it("shouldn't warn if dupewarn is false", async () => {
                    const spy = jest.spyOn(global.console, "warn");

                    spy.mockImplementation(() => { /* NO-OP */ });

                    const processor = new Processor({
                        namer,
                        dupewarn : false,
                    });

                    await processor.string("packages/processor/test/specimens/start.css", ".start { color: red; }");
                    await processor.string("packages/processor/test/specimens/START.css", ".start { color: red; }");

                    expect(spy).not.toHaveBeenCalled();
                });
            });
        });
    });
});
