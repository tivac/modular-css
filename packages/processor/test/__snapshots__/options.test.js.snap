// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js options exportGlobals should not export :global values when exportGlobals is false 1`] = `
Object {
  "b": Array [
    "mc8df92ddf_b",
  ],
}
`;

exports[`/processor.js options lifecycle options after should run async postcss plugins 1`] = `
"/* packages/processor/test/specimens/relative.css */
.mc_wooga {
    color: red;
    background: url(\\"../specimens/folder/to.png\\");
}
a {}
"
`;

exports[`/processor.js options lifecycle options after should run sync postcss plugins 1`] = `
"/* packages/processor/test/specimens/relative.css */
.mc_wooga {
    color: red;
    background: url(\\"../specimens/folder/to.png\\");
}
a {}
"
`;

exports[`/processor.js options lifecycle options after should use postcss-url by default 1`] = `
"/* packages/processor/test/specimens/relative.css */
.mc8e3ffde3_wooga {
    color: red;
    background: url(\\"../specimens/folder/to.png\\");
}
"
`;

exports[`/processor.js options lifecycle options before should run async postcss plugins before processing 1`] = `
"/* packages/processor/test/specimens/async-before.css */
a {}"
`;

exports[`/processor.js options lifecycle options before should run sync postcss plugins before processing 1`] = `
"/* packages/processor/test/specimens/sync-before.css */
a {}"
`;

exports[`/processor.js options lifecycle options done should run async postcss plugins done processing 1`] = `
"/* packages/processor/test/specimens/async-done.css */
a {}"
`;

exports[`/processor.js options lifecycle options done should run sync postcss plugins done processing 1`] = `
"/* packages/processor/test/specimens/sync-done.css */
a {}"
`;

exports[`/processor.js options lifecycle options dupewarn should warn on potentially duplicate file paths 1`] = `
Array [
  Array [
    "POTENTIAL DUPLICATE FILES:
	packages/processor/test/specimens/start.css
	packages/processor/test/specimens/START.css",
  ],
]
`;

exports[`/processor.js options lifecycle options processing should include exports from 'modular-css-export' modules 1`] = `
Object {
  "a": true,
  "b": false,
}
`;

exports[`/processor.js options lifecycle options processing should run async postcss plugins processing processing 1`] = `
"/* packages/processor/test/specimens/async-processing.css */
a {}"
`;

exports[`/processor.js options lifecycle options processing should run sync postcss plugins processing processing 1`] = `
"/* packages/processor/test/specimens/sync-processing.css */
a {}"
`;

exports[`/processor.js options lifecycle options verbose should output debugging messages when verbose mode is enabled 1`] = `
Array [
  Array [
    "[processor]",
    "file()",
    "packages/processor/test/specimens/start.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/processor/test/specimens/start.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/processor/test/specimens/start.css",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/processor/test/specimens/local.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/processor/test/specimens/local.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/processor/test/specimens/local.css",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/processor/test/specimens/folder/folder.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/processor/test/specimens/folder/folder.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/processor/test/specimens/folder/folder.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/processor/test/specimens/folder/folder.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/processor/test/specimens/local.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/processor/test/specimens/start.css",
  ],
  Array [
    "[processor]",
    "string()",
    "packages/processor/test/specimens/string.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/processor/test/specimens/string.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/processor/test/specimens/string.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/processor/test/specimens/string.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/processor/test/specimens/folder/folder.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/processor/test/specimens/string.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/processor/test/specimens/local.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/processor/test/specimens/start.css",
  ],
]
`;

exports[`/processor.js options map should generate external source maps 1`] = `
"/* packages/processor/test/specimens/folder/folder.css */
.mc_folder { margin: 2px; }
/* packages/processor/test/specimens/local.css */
.mc_booga { background: green; }
/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInBhY2thZ2VzL3Byb2Nlc3Nvci90ZXN0L3NwZWNpbWVucy9mb2xkZXIvZm9sZGVyLmNzcyIsInBhY2thZ2VzL3Byb2Nlc3Nvci90ZXN0L3NwZWNpbWVucy9sb2NhbC5jc3MiLCJwYWNrYWdlcy9wcm9jZXNzb3IvdGVzdC9zcGVjaW1lbnMvc3RhcnQuY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLHdEQUFBO0FBQ0EsYUFBVSxXQUFXLEVBQUU7QUNEdkIsZ0RBQUE7QUFJQSxZQUFTLGlCQUFpQixFQUFFO0FDSjVCLGdEQUFBO0FBRUEsWURGQSxVQUFlLEVBQ2YsZ0JBQWdCLEVDQ3NCO0FBQ3RDLFlGSEEsdUJBQW9CLEVFR2UiLCJmaWxlIjoib3V0LmNzcyIsInNvdXJjZXNDb250ZW50IjpbIkB2YWx1ZSBmb2xkZXI6IHdoaXRlO1xuLmZvbGRlciB7IG1hcmdpbjogMnB4OyB9XG4iLCJAdmFsdWUgb25lOiByZWQ7XG5AdmFsdWUgdHdvOiBibHVlO1xuQHZhbHVlIGZvbGRlciBmcm9tIFwiLi9mb2xkZXIvZm9sZGVyLmNzc1wiO1xuXG4uYm9vZ2EgeyBiYWNrZ3JvdW5kOiBncmVlbjsgfVxuLmxvb2dhIHsgY29tcG9zZXM6IGJvb2dhOyB9XG4iLCJAdmFsdWUgb25lLCB0d28sIGZvbGRlciBmcm9tIFwiLi9sb2NhbC5jc3NcIjtcbi53b29nYSB7IGNvbXBvc2VzOiBib29nYSBmcm9tIFwiLi9sb2NhbC5jc3NcIjsgfVxuLmJvb2dhIHsgY29sb3I6IG9uZTsgYmFja2dyb3VuZDogdHdvOyB9XG4udG9vZ2EgeyBib3JkZXI6IDFweCBzb2xpZCBmb2xkZXI7IH1cbiJdfQ== */"
`;

exports[`/processor.js options map should generate source maps 1`] = `
"/* packages/processor/test/specimens/folder/folder.css */
.mc_folder { margin: 2px; }
/* packages/processor/test/specimens/local.css */
.mc_booga { background: green; }
/* packages/processor/test/specimens/start.css */
.mc_booga { color: red; background: blue; }
.mc_tooga { border: 1px solid white; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInBhY2thZ2VzL3Byb2Nlc3Nvci90ZXN0L3NwZWNpbWVucy9mb2xkZXIvZm9sZGVyLmNzcyIsInBhY2thZ2VzL3Byb2Nlc3Nvci90ZXN0L3NwZWNpbWVucy9sb2NhbC5jc3MiLCJwYWNrYWdlcy9wcm9jZXNzb3IvdGVzdC9zcGVjaW1lbnMvc3RhcnQuY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLHdEQUFBO0FBQ0EsYUFBVSxXQUFXLEVBQUU7QUNEdkIsZ0RBQUE7QUFJQSxZQUFTLGlCQUFpQixFQUFFO0FDSjVCLGdEQUFBO0FBRUEsWURGQSxVQUFlLEVBQ2YsZ0JBQWdCLEVDQ3NCO0FBQ3RDLFlGSEEsdUJBQW9CLEVFR2UiLCJmaWxlIjoib3V0LmNzcyIsInNvdXJjZXNDb250ZW50IjpbIkB2YWx1ZSBmb2xkZXI6IHdoaXRlO1xuLmZvbGRlciB7IG1hcmdpbjogMnB4OyB9XG4iLCJAdmFsdWUgb25lOiByZWQ7XG5AdmFsdWUgdHdvOiBibHVlO1xuQHZhbHVlIGZvbGRlciBmcm9tIFwiLi9mb2xkZXIvZm9sZGVyLmNzc1wiO1xuXG4uYm9vZ2EgeyBiYWNrZ3JvdW5kOiBncmVlbjsgfVxuLmxvb2dhIHsgY29tcG9zZXM6IGJvb2dhOyB9XG4iLCJAdmFsdWUgb25lLCB0d28sIGZvbGRlciBmcm9tIFwiLi9sb2NhbC5jc3NcIjtcbi53b29nYSB7IGNvbXBvc2VzOiBib29nYSBmcm9tIFwiLi9sb2NhbC5jc3NcIjsgfVxuLmJvb2dhIHsgY29sb3I6IG9uZTsgYmFja2dyb3VuZDogdHdvOyB9XG4udG9vZ2EgeyBib3JkZXI6IDFweCBzb2xpZCBmb2xkZXI7IH1cbiJdfQ== */"
`;

exports[`/processor.js options namer should require a namer if a string is passed 1`] = `
Object {
  "wooga": Array [
    "AA",
  ],
}
`;

exports[`/processor.js options namer should use a custom naming function 1`] = `
Object {
  "wooga": Array [
    "packages_processor_test_specimens_simple_css_wooga",
  ],
}
`;

exports[`/processor.js options namer should use a custom naming function 2`] = `
Object {
  "wooga": Array [
    "packages_processor_test_specimens_simple_css_wooga",
  ],
}
`;

exports[`/processor.js options namer should use a custom naming function 3`] = `".packages_processor_test_specimens_simple_css_wooga { }"`;

exports[`/processor.js options namer should use the default naming function if a non-function is passed 1`] = `
Object {
  "wooga": Array [
    "mc4a20438a_wooga",
  ],
}
`;

exports[`/processor.js options postcss options should support custom parsers 1`] = `
"/* packages/processor/test/specimens/parser.css */
.mce1d18d5d_a {
    color: blue
}"
`;

exports[`/processor.js options rewrite should not rewrite url() references when falsey 1`] = `
"/* packages/processor/test/specimens/rewrite.css */
.mc20ca2799_a {
    background: url(\\"img.png\\");
}"
`;

exports[`/processor.js options rewrite should pass through to postcss-url as config 1`] = `
"/* packages/processor/test/specimens/rewrite.css */
.mc20ca2799_a {
    background: url(\\"data:image/png;base64,iVBORwoaCgAAAA1JSERSAAAACgAAAAoIAgAAAAJQWOoAAAABc1JHQgCuzhzpAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwwAADsMBx2+oZAAAABl0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC4xOdTWsmQAAAAOSURBVChTYxgFJAMGBgABNgABY8OiGAAAAABJRU5ErkJggg==\\");
}"
`;

exports[`/processor.js options rewrite should rewrite url() references by default 1`] = `
"/* packages/processor/test/specimens/rewrite.css */
.mc20ca2799_a {
    background: url(\\"../specimens/img.png\\");
}"
`;
